internal/pkg/agent/application/dispatcher/dispatcher.go

```diff
@@ -192,70 +203,165 @@ func detectTypes(actions []fleetapi.Action) []string {
+		len(dequeuedActions), len(expired), len(expiredUpgradeActions))
+	ad.log.Debugf("Expired non-upgrade actions: %v", expired)
+	ad.log.Debugf("Expired upgrade actions (won't be removed from the queue): %v", expiredUpgradeActions)
+
+	for _, expiredUpgradeAction := range expiredUpgradeActions {
+		startTime, err := expiredUpgradeAction.StartTime()
+		if err != nil {
+			ad.log.Warnf("failed to get start time of expired upgrade action [%s]: %v", expiredUpgradeAction.ID(), err)
 			continue
 		}
-		queued = append(queued, action)
+		ad.queue.Add(expiredUpgradeAction, startTime.Unix())
+	}
+
+	// if an upgrade action is included in the immediate dispatchable actions
+	// mark upgradeDetailsNeedUpdate as false since the upgrade details will be set by it
+	if slices.ContainsFunc(actions, func(action fleetapi.Action) bool {
+		return action.Type() == fleetapi.ActionTypeUpgrade
+	}) {
+		*upgradeDetailsNeedUpdate = false
 	}
-	return queued, expired
+
+	return actions
 }
 
-// removeQueuedUpgrades will scan the passed actions and if there is an upgrade action it will remove all upgrade actions in the queue but not alter the passed list.
-// this is done to try to only have the most recent upgrade action executed. However it does not eliminate duplicates in retrieved directly from the gateway
-func (ad *ActionDispatcher) removeQueuedUpgrades(actions []fleetapi.Action) {
+// removeQueuedUpgrades will scan the passed actions and if there is an upgrade action it will remove all upgrade actions
+// in the queue but not alter the passed list. This is done to try to only have the most recent upgrade action executed.
+// However, it does not eliminate duplicates in retrieved directly from the gateway. Also, if upgrade actions are removed
+// from the queue, upgradeDetailsNeedUpdate will be set to true.
+func (ad *ActionDispatcher) removeQueuedUpgrades(actions []fleetapi.Action, upgradeDetailsNeedUpdate *bool) {
 	for _, action := range actions {
 		if action.Type() == fleetapi.ActionTypeUpgrade {
 			if n := ad.queue.CancelType(fleetapi.ActionTypeUpgrade); n > 0 {
 				ad.log.Debugw("New upgrade action retrieved from gateway, removing queued upgrade actions", "actions_found", n)
+				// upgrade action(s) got removed from the queue so upgrade actions changed
+				*upgradeDetailsNeedUpdate = true
 			}
 			return
 		}
 	}
 }
 
+// updateUpgradeDetails will construct the upgrade details based queue actions (assuming expired ones are still in the queue)
+// and if ad.lastUpgradeDetails is different from the new upgrade details, it will update them.
+func (ad *ActionDispatcher) updateUpgradeDetails(ts time.Time, detailsSetter details.Observer) {
+	// no upgrade details from expired actions check the stored actions in the queue
+	upgradeDetails := GetScheduledUpgradeDetails(ad.log, ad.queue.Actions(), ts)
+
+	if ad.lastUpgradeDetailsIsSet && ad.lastUpgradeDetails.Equals(upgradeDetails) {
+		// calculate
```

Comment:
```suggestion
		// already the same details; do nothing
```

Reply:
changed in 6f75b29b772af12fe38091555902a056c14b9f35